### PR TITLE
BoxResolver impl

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: spectrum-offchain/src/box_resolver/process.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    rollbacks: UnboundedReceiver<UpgradeRollback<TEntity>>,
+) -> impl Stream<Item = ()> + 'a
+where
+    TRepo: EntityRepo<TEntity, TEntityId, TStateId> + Clone + 'a,
+    TEntity: Has<TEntityId> + Has<TStateId> + 'a,
+{
+    select_all(vec![
+        track_conf_upgrades(persistence.clone(), conf_upgrades),
+        track_unconf_upgrades(persistence.clone(), unconf_upgrades),
+        handle_rollbacks(persistence.clone(), rollbacks),
```
Comment (kettlebell) [inline spectrum-offchain/src/box_resolver/process.rs:25]: Redundant clone

File: spectrum-offchain/src/data/reprod_entity/state.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/// A unique, persistent, self-reproducible, on-chiain entity.
```
Comment (kettlebell) [inline spectrum-offchain/src/data/reprod_entity/state.rs:1]: "...on-chain entity."

File: spectrum-offchain/src/event_sink.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    TUpstream: Stream<Item = TEvent> + 'a,
+    TEvent: Clone + 'a,
+    TDefHan: DefaultEventHandler<TEvent> + 'a,
+{
+    upstream
+        .for_each(move |ev| {
+            let mut unhandled_ev = None;
+            for han in &mut handlers {
+                let maybe_unhandled_ev = han.try_handle(ev.clone());
+                if unhandled_ev.is_some() {
```
Comment (kettlebell) [inline spectrum-offchain/src/event_sink.rs:23]: should be `is_none()` instead?

File: spectrum-offchain/src/event_sink/types.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+pub trait EventHandler<TEvent> {
+    /// Tries to handle the gicen event if applicable.
```
Comment (kettlebell) [inline spectrum-offchain/src/event_sink/types.rs:2]: "...given event..."

File: ergo-chain-sync/src/lib.rs [hunk 5]
```diff
@@ -102,16 +127,16 @@ impl<TClient, TCache> ChainSync<TClient, TCache>
...
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        let mut fut = Box::pin(self.try_upgrade());
+        let mut upgr_fut = Box::pin(self.try_upgrade());
         loop {
-            match fut.as_mut().poll(cx) {
+            match upgr_fut.as_mut().poll(cx) {
                 Poll::Ready(Some(upgr)) => return Poll::Ready(Some(upgr)),
                 Poll::Ready(None) => return Poll::Pending,
                 Poll::Pending => continue,
```
[the remark this answers is not in the document]
Comment (oskin1) [reply ergo-chain-sync/src/lib.rs:142]: I guess if we just return pending here, then `upgr_fut` will be dropped and won't get c chance to progress. Next call of `poll` will spawn a new future. We should memoize `upgr_fut` if we don't want to spin.

File: ergo-mempool-sync/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                    Poll::Ready(None) => {}
+                    Poll::Pending => break,
+                }
+            }
+            let mut sync_fut = Box::pin(sync(this.client, state));
+            // Drive sync to completion
+            loop {
+                match sync_fut.as_mut().poll(cx) {
+                    Poll::Ready(_) => break,
+                    Poll::Pending => continue,
```
[the remark this answers is not in the document]
Comment (oskin1) [reply ergo-mempool-sync/src/lib.rs:157]: Same thing as above, not sure what's the right solution here. Please share your thoughts.

File: spectrum-offchain/src/box_resolver/process.rs (hunk 1 above)
Comment (oskin1) [reply spectrum-offchain/src/box_resolver/process.rs:25]: This piece is rewritten in the subsequent PR

File: spectrum-offchain/src/event_sink.rs (hunk 3 above)
Comment (oskin1) [reply spectrum-offchain/src/event_sink.rs:23]: Fixed.

File: ergo-mempool-sync/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergo-mempool-sync/src/lib.rs:157]: As you mentioned in slack, if we slow down the polling frequency, it allows others to speed up their polling which will confer an advantage for order execution.

File: ergo-chain-sync/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergo-chain-sync/src/lib.rs:142]: If we get back `Poll::Ready(None)` then return `Poll:Pending`, no waker will have been set and so this stream will stall on the executor. In light of the race for order execution, it's probably better to spin for this case too.

[the remark this answers is not in the document]
Comment (kettlebell) [reply ergo-chain-sync/src/lib.rs:142]: Yes.

**Pull request merged**